### ISSUE TITLE
Not sure how to get a basic test working

### ISSUE THREAD

**Issue by Youdaman** [excid3/revise_auth#80]
I'm fairly new to Ruby and Rails so no doubt I've done something wrong, however I can't work out how to use the login helper.

I've set things up per as the readme, however I did also have to add the following require to test_helper.rb to avoid `uninitialized constant ReviseAuth::Test (NameError)`:
```ruby
require 'C:\Ruby33-x64\lib\ruby\gems\3.3.0\gems\revise_auth-0.6.0\lib\revise_auth\test\helpers.rb'
```

Here's a test to just show the index for a controller (which requires authentication):
```ruby
class GamesControllerTest < ActionDispatch::IntegrationTest
  setup do
    @user = users(:bob)
    login(@user, password: "password")
    @game = games(:one)
  end

  test "should get index" do
    get games_url
    assert_response :success
  end
```
Running that test by itself it fails and I don't know why:
```
❯ bin/rails test test/controllers/games_controller_test.rb:10
Running 7 tests in a single process (parallelization threshold is 50)
Run options: --seed 34325

# Running:

F

Failure:
GamesControllerTest#test_should_get_index [test/controllers/games_controller_test.rb:13]:
Expected response to be a <2XX: success>, but was a <403: Forbidden>


bin/rails test test/controllers/games_controller_test.rb:10



Finished in 1.262554s, 0.7920 runs/s, 0.7920 assertions/s.
1 runs, 1 assertions, 1 failures, 0 errors, 0 skips
```

The user bob exists in the fixtures (users.yml):
```
bob:
  email: bob@bob.com
  password_digest: <%= BCrypt::Password.create("password") %>
```

Any clues would be appreciated!

**Comment by gathuku** [comment]
Instead of the custom require you can open `ActionDispatch::IntegrationTest` class and include the test helpers.

```rb
# test_helper.rb
class ActionDispatch::IntegrationTest
  include ReviseAuth::Test::Helpers
end
```
Not sure why you're getting a 403 error.

**Comment by Youdaman** [comment]
Thanks @gathuku but I already had the test helpers included, here's my test_helper.rb:
```ruby
ENV["RAILS_ENV"] ||= "test"
require_relative "../config/environment"
require "rails/test_help"
require 'C:\Ruby33-x64\lib\ruby\gems\3.3.0\gems\revise_auth-0.6.0\lib\revise_auth\test\helpers.rb'

# needed for password_digest in users.yml fixtures
require "bcrypt"

module ActiveSupport
  class TestCase
    # Run tests in parallel with specified workers
    parallelize(workers: :number_of_processors, with: :threads)

    # Setup all fixtures in test/fixtures/*.yml for all tests in alphabetical order.
    fixtures :all

    # Add more helper methods to be used by all tests here...
  end
end

# see https://github.com/excid3/revise_auth?tab=readme-ov-file#test-helpers
class ActionDispatch::IntegrationTest
  include ReviseAuth::Test::Helpers
end
```

This was the default test_helper.rb with the only parts added being the revise test helpers class and require, and the require for bycrypt (for the password_digest fields of the user fixtures).

I'm wondering if it's anything to do with being on Windows?

Am I correct in understanding that tests by default don't autoload modules like the app does? In which case this is what makes the explicit require for the revise test helpers necessary?

**Comment by gathuku** [comment]
@Youdaman I think rails autoload all gems in your Gemfile. It's supposed to work without the explicit require. 

Are you raising forbidden error anywhere in your `GamesController`? or how does the code looks like.

**Comment by Youdaman** [comment]
My guess is that I haven't started with tests until now and I've been working on this project for a while now, so there's likely some change I've made to the config or another file that is breaking things.

I just created a fresh project, added revise_auth, added the helpers to the test and ran `bin/rails test` with no problems. I'll start my project again but this time run tests from the very beginning so that whenever I make a change I can run the tests to see what breaks things.

Apologies for the waste of time, closing this!

**Comment by Youdaman** [comment]
@gathuku Update -- it turns out that I was using revise 0.5... and when I updated to 0.7 things worked... so that helper class wasn't found because it mustn't have been in 0.5! Tests run now! 😄

**Comment by Youdaman** [comment]
I stripped down the users fixture in case the confirmed/email fields were breaking things but still no luck:
```yml
one:
  email: one@example.com
  password_digest: <%= BCrypt::Password.create("password") %>

two:
  email: two@example.com
  password_digest: <%= BCrypt::Password.create("password") %>
```

**Comment by gathuku** [comment]
@yorickvandervis not entirely sure by instead of using `password_digest` try `password` in fixtures and supply a plain text.

**Comment by gathuku** [comment]
@yorickvandervis added this https://github.com/gathuku/revise_auth_test/commit/ff04b4dcda5cddca663afd4cca6c5e525cd7c01d and seems to work fine. 

I think you have other part of your application logic you need to check

**Comment by Youdaman** [comment]
Thanks @gathuku I think you're right about it being another part of my application logic interfering because the test fails with:
```
❯ bin/rails test test/controllers/test_controller_test.rb 
Running 2 tests in a single process (parallelization threshold is 50)
Run options: --seed 25006

# Running:

F

Failure:
TestControllerTest#test_index:_success [test/controllers/test_controller_test.rb:9]:
Expected response to be a <200: ok>, but was a <403: Forbidden>.
Expected: 200
  Actual: 403


bin/rails test test/controllers/test_controller_test.rb:4

F

Failure:
TestControllerTest#test_index:_failure [test/controllers/test_controller_test.rb:15]:
Expected response to be a <3XX: redirect>, but was a <403: Forbidden>


bin/rails test test/controllers/test_controller_test.rb:12



Finished in 1.177078s, 1.6991 runs/s, 1.6991 assertions/s.
2 runs, 2 assertions, 2 failures, 0 errors, 0 skips
```
Leave it with me and I'll work it out, even if I have to start a fresh project. Thanks again for your assistance!

**Comment by Youdaman** [comment]
@gathuku hey Moses, I found the issue, you were right it was caused by something else. Turns out I'd set the `config.hosts` and not included "www.example.com" so the tests wouldn't work when trying to hit that domain. I found out by running the debugger in the tests and checking what the request variable was for your successful example versus my failing tests, and noticed yours had various things like auth cookies whereas mine just had HTML and the page said "Blocked hosts: www.example.com" which gave me the clue :) Thanks again for trying to help me!
